crypto/fipsmodule/aes/asm/aesni-avx512.pl

```diff
@@ -0,0 +1,3165 @@
+# VAES and VPCLMULQDQ instructions with full width of ZMM registers.
+
+# The first two arguments should always be the flavour and output file path.
+if ($#ARGV < 1) { die "Not enough arguments provided.
+  Two arguments are necessary: the flavour and the output file path."; }
+
+$flavour = shift;
+$output  = shift;
+
+$win64=0; $win64=1 if ($flavour =~ /[nm]asm|mingw64/ || $output =~ /\.asm$/);
+
+$avx512vaes = 1;
+for (@ARGV) { $avx512vaes = 0 if (/-DMY_ASSEMBLER_IS_TOO_OLD_FOR_512AVX/); }
+
+$0 =~ m/(.*[\/\\])[^\/\\]+$/; $dir=$1;
+( $xlate="${dir}x86_64-xlate.pl" and -f $xlate ) or
+( $xlate="${dir}../../../perlasm/x86_64-xlate.pl" and -f $xlate) or
+die "can't locate x86_64-xlate.pl";
+
+open OUT,"| \"$^X\" \"$xlate\" $flavour \"$output\"";
+*STDOUT=*OUT;
+
+#======================================================================
+
+if ($avx512vaes) {
+
+  my $GP_STORAGE  = $win64 ? (16 * 33)  : (16 * 23);    # store rbx
+  my $XMM_STORAGE = $win64 ? (16 * 23) : 0;     # store xmm6:xmm15
+  my $VARIABLE_OFFSET = $win64 ? (16 *8 + 16* 15 + 16 * 10 + 8*3) :
+                                 (16*8 + 16 * 15 + 8 * 1);
+
+  my $TW = "%rsp";
+  my $TWTEMPH = "%rbx";
+  my $TWTEMPL = "%rax";
+  my $ZPOLY = "%zmm25";
+
+  # ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+  # ;;; Function arguments abstraction
+  # ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+  my ($key2, $key1, $tweak, $length, $input, $output);
+
+  if ($win64) {
+    $input    = "%rcx";
+    $output   = "%rdx";
+    $length   = "%r8";
+    $key1     = "%r9";
+    $key2     = "%r10";
+    $tweak    = "%r11";
+  } else {
+    $input    = "%rdi";
+    $output   = "%rsi";
+    $length   = "%rdx";
+    $key1     = "%rcx";
+    $key2     = "%r8";
+    $tweak    = "%r9";
+  }
+
+  # arguments for temp parameters
+  my ($tmp1, $ghash_poly_8b, $ghash_poly_8b_temp);
+  if ($win64) {
+    $tmp1                = "%r10";
+    $ghash_poly_8b       = "%rdi";
+    $ghash_poly_8b_temp  = "%rsi";
+  } else {
+    $tmp1                = "%r8";
+    $ghash_poly_8b       = "%r10";
+    $ghash_poly_8b_temp  = "%r11";
+  }
+
+  # ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+  # ;;; Helper functions
+  # ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+  # Generates "random" local labels
+  sub random_string() {
+    my @chars  = ('a' .. 'z', 'A' .. 'Z', '0' .. '9', '_');
+    my $length = 15;
+    my $str;
+    map { $str .= $chars[rand(33)] } 1 .. $length;
+    return $str;
+  }
+
+  # ; Seed the RNG so the labels are generated deterministically
+  srand(12345);
+
+  sub encrypt_tweak_for_encryption {
+    my $key2               = $_[0];
+    my $state_tweak        = $_[1];
+    my $key1               = $_[2];
+    my $raw_key            = $_[3];
+    my $tmp                = $_[4];
+    my $ptr_key2           = $_[5];
+    my $ptr_key1           = $_[6];
+    my $ptr_expanded_keys  = $_[7];
+
+    $code.=<<___;
+    vmovdqu  ($ptr_key2), $key2
+    vpxor    $key2, $state_tweak, $state_tweak  # ARK for tweak encryption
```

Comment:
Suggestion for first occurrence of ARK:
```suggestion
    vpxor    $key2, $state_tweak, $state_tweak  # AddRoundKey (ARK) for tweak encryption
```